echoip: Init at unstable-2018-11-20

rvolosatovs opened the pull request:
###### Motivation for this change
The power behind https://ifconfig.co now in `nixpkgs` :tada: 
~The project uses go modules - I did not find an automated solution to convert the go module files into `deps.nix`, so I first converted it to use `dep`, then used `dep2nix` to generate the `deps.nix`.
Still had to do some manual tweaking to make it build, however, as `golang.org/x/sys` was not included in `deps.nix` by some reason.~
Turns out https://github.com/adisbladis/vgo2nix exists, so I just used the tool to generate `deps.nix`

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

rvolosatovs committed:
echoip: Init at unstable-2018-11-20

worldofpeace commented on pkgs/applications/misc/echoip/default.nix:
```diff
@@ -0,0 +1,24 @@
+
+buildGoPackage rec {
+  name = "echoip-${version}";
+  version = "20181120-${lib.strings.substring 0 7 rev}";
```
This should be  `unstable-2018-11-20`

> If a package is not a release but a commit from a repository, then the version part of the name must be the date of that (fetched) commit. The date must be in "YYYY-MM-DD" format. Also append "unstable" to the name - e.g., "pkgname-unstable-2014-09-23". 

##### https://nixos.org/nixpkgs/manual/#sec-package-naming

rvolosatovs commented:
@worldofpeace 
- fixed version
- moved to `servers`
- added `index.html` to `out`

Example usage: https://github.com/rvolosatovs/infrastructure/blob/fae0779f897a527a628f5faeca88bcb61df12be4/nixos/echoip.nix#L17

rvolosatovs commented:
...And fixed a typo

worldofpeace commented:
Corrected the hash manually just to test with https://github.com/rvolosatovs/infrastructure/blob/fae0779f897a527a628f5faeca88bcb61df12be4/nixos/echoip.nix#L17 and it does work.

A NixOS module would be nice too, but it's not required.

rvolosatovs commented:
```bash
rg 'fetchSubmodules|submodule' $(find $(nix eval '(<nixpkgs>)') -name 'deps.nix')
```

Does not find anything either, is it not possible to control the submodule fetching in go package builds?

worldofpeace commented:
@rvolosatovs 

Looking at the `buildGoPackage` source, this looks pertinent:

```nix
dep2src = goDep:
    {
      inherit (goDep) goPackagePath;
      src = if goDep.fetch.type == "git" then
        fetchgit {
          inherit (goDep.fetch) url rev sha256;
        }
...
```

It doesn't care about `fetchSubmodules = false;`. That's probably something we should fix in the future.

If you want to you can work around this by changing the fetch type to `FromGitHub` which will use a github archive without any submodules.

```nix
 {
    goPackagePath  = "github.com/oschwald/geoip2-golang";
    fetch = {
      type = "FromGitHub";
      owner = "oschwald";
      repo = "geoip2-golang";
      rev =  "7118115686e16b77967cdbf55d1b944fe14ad312";
      sha256 = "0zpgpz577rghvgis6ji9l99pq87z5izbgzmnbyn3dy533bayrgpw";
    };
  }
```

(hash retrieved from `nix-prefetch-github`)

rvolosatovs commented:
@worldofpeace implemented the workaround

worldofpeace commented on pkgs/servers/echoip/default.nix:
```diff
@@ -0,0 +1,30 @@
+    sha256 = "0n5d9i8cc5lqgy5apqd3zhyl3h1xjacf612z8xpvbm75jnllcvxy";
+  };
+
+  goDeps = ./deps.nix;
```
Can you add a comment here stating that this file needs special care? Mostly for the reasons outlined in the OP.

worldofpeace commented:
Still getting a hash issue so I think the `FetchGitHub` workaround is still needed.

```
fixed-output derivation produced path '/nix/store/g2x3lfhppx3g9lwrffjxkjwal7cz51ki-geoip2-golang' with sha256 hash '0d3kykb1xqh4a1qkpsbl7bbk1pcffshj2dvp5b3xrhyvxfs2cdxs' instead of the expected hash '0zpgpz577rghvgis6ji9l99pq87z5izbgzmnbyn3dy533bayrgpw'
```
